receiver/azuremonitorreceiver/scraper.go

```diff
@@ -0,0 +1,339 @@
+		metadata.WithAzureMonitorSubscriptionID(s.cfg.SubscriptionId),
+		metadata.WithAzureMonitorTenantID(s.cfg.TenantId),
+	), nil
+}
+
+func (s *azureScraper) getResources(ctx context.Context) {
+	if time.Now().UTC().Unix() < (s.resourcesUpdated + s.cfg.CacheResources) {
+		return
+	}
+	existingResources := map[string]void{}
+	for id := range s.resources {
+		existingResources[id] = void{}
+	}
+
+	filter := s.getResourcesFilter()
+	opts := &armresources.ClientListOptions{
+		Filter: &filter,
+	}
+
+	pager := s.clientResources.NewListPager(opts)
+
+	for pager.More() {
+		nextResult, err := pager.NextPage(ctx)
+		if err != nil {
+			s.settings.Logger.Error("failed to get Azure Resources data", zap.Error(err))
+			return
+		}
+		for _, resource := range nextResult.Value {
+
+			if _, ok := s.resources[*resource.ID]; !ok {
+				s.resources[*resource.ID] = &azureResource{}
+			}
+			delete(existingResources, *resource.ID)
+		}
+	}
+	if len(existingResources) > 0 {
+		for idToDelete := range existingResources {
+			delete(s.resources, idToDelete)
+		}
+	}
+
+	s.resourcesUpdated = time.Now().UTC().Unix()
+}
+
+func (s *azureScraper) getResourcesFilter() string {
+	// TODO: switch to parsing services from https://learn.microsoft.com/en-us/azure/azure-monitor/essentials/metrics-supported
+	var resourcesTypeFilter string
+	if len(s.cfg.Services) > 0 {
+		resourcesTypeFilter = strings.Join(s.cfg.Services, "' or resourceType eq '")
+	} else {
+		resourcesTypeFilter = strings.Join(monitorServices, "' or resourceType eq '")
+	}
+
+	resourcesGroupFilterString := ""
+	if len(s.cfg.ResourceGroups) > 0 {
+		resourcesGroupFilterString = fmt.Sprintf(" and (resourceGroup eq '%s')",
+			strings.Join(s.cfg.ResourceGroups, "' or resourceGroup eq  '"))
+	}
+
+	return fmt.Sprintf("(resourceType eq '%s')%s", resourcesTypeFilter, resourcesGroupFilterString)
+}
+
+func (s *azureScraper) getResourceMetricsDefinitions(ctx context.Context, resourceId string) {
+
+	if time.Now().UTC().Unix() < (s.resources[resourceId].metricsDefinitionsUpdated + s.cfg.CacheResourcesDefinitions) {
+		return
+	}
+	res := s.resources[resourceId]
+	res.metricsByGrains = map[string]*azureResourceMetrics{}
+
+	pager := s.clientMetricsDefinitions.NewListPager(resourceId, nil)
+	for pager.More() {
+		nextResult, err := pager.NextPage(ctx)
+		if err != nil {
+			s.settings.Logger.Error("failed to get Azure Metrics definitions data", zap.Error(err))
+			return
+		}
+		for _, v := range nextResult.Value {
+
+			timeGrain := *v.MetricAvailabilities[0].TimeGrain
+			name := *v.Name.Value
+
+			if _, ok := res.metricsByGrains[timeGrain]; ok {
+				res.metricsByGrains[timeGrain].metrics = append(res.metricsByGrains[timeGrain].metrics, name)
+			} else {
+				res.metricsByGrains[timeGrain] = &azureResourceMetrics{metrics: []string{name}}
+			}
+		}
+	}
+	res.metricsDefinitionsUpdated = time.Now().UTC().Unix()
+}
+
+func (s *azureScraper) getResourceMetricsValues(ctx context.Context, resourceId string) {
```

Comment:
Usually 2 levels of iteration is about the limit of what I can comprehend after 2 pm. So I end up virtually (and sometimes literally) rewriting functions like this one when I need to understand some code. Could you extract at least one or two helpers separate operations?